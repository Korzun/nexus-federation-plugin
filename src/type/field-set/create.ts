import { TypeCreate } from '../../builder';

import { model, Options as ModelOptions } from './model';

export type Options = ModelOptions;

export const create: TypeCreate<Options> = (builder, options) => {
  builder.addOnInstallFunction((t) => {
    t.addType(model(options));
  });

  return builder;
};
