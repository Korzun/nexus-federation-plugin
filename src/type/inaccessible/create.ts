import { addDirective, core } from 'nexus';

import { TypeCreate } from '../../builder';

import { directive } from './directive';
import { field } from './field';
import { inputObject } from './input-object';
import { object } from './object';

//TODO: Add missing type for argType, enum, enumValue
export const create: TypeCreate = (builder) => {
  builder.addOnInstallFunction((t) => {
    t.addType(directive);
  });

  /**
   * Adds `inaccessible` to Nexus field definitions and generates the
   * `@inaccessible` directive when it is detected in a field definition
   */
  builder.addFieldDefType(field);
  builder.addOnAddFieldFunction((fieldConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (fieldConfig.inaccessible) {
      if (!fieldConfig.directives) {
        fieldConfig.directives = new Array();
      }
      (fieldConfig.directives as Array<core.NexusDirectiveUse>).push(
        addDirective('inaccessible', {}),
      );
    }
  });

  /**
   * Adds `inaccessible` to Nexus object definitions and generates the
   * `@inaccessible` directive when it is detected in an object definition
   */
  builder.addObjectDefType(object);
  builder.addOnObjectDefinitionFunction((t, objectConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (objectConfig.inaccessible) {
      if (!objectConfig.directives) {
        objectConfig.directives = new Array();
      }
      (objectConfig.directives as Array<core.NexusDirectiveUse>).push(
        addDirective('inaccessible', {}),
      );
    }
  });

  /**
   * Adds `inaccessible` to Nexus input field definitions and generates the
   * `@inaccessible` directive when it is detected in an input field definition.
   *
   * For some reason it is not necessary to call `addInputFieldDefType` to add
   * the input field. The type added through `addFieldDefType` appears to be sufficient
   */
  builder.addOnAddInputFieldFunction((inputFieldConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (inputFieldConfig.inaccessible) {
      if (!inputFieldConfig.directives) {
        inputFieldConfig.directives = new Array();
      }
      (inputFieldConfig.directives as Array<core.NexusDirectiveUse>).push(
        addDirective('inaccessible', {}),
      );
    }
  });

  //FIXME: Nexus Bug - the key `inaccessible` isn't properly surfaced in onAdd function
  // /**
  //  * Adds `inaccessible` to Nexus input object definitions and generates the
  //  * `@inaccessible` directive when it is detected in an input object definition
  //  */
  builder.addInputObjectDefType(inputObject);
  // builder.addOnInputObjectDefinitionFunction((inputObjectConfig) => {
  //   if (inputObjectConfig.inaccessible) {
  //     if (!inputObjectConfig.directives) {
  //       inputObjectConfig.directives = new Array();
  //     }
  //     (inputObjectConfig.directives as Array<core.NexusDirectiveUse>).push(
  //       addDirective('inaccessible', {}),
  //     );
  //   }
  // });

  return builder;
};
