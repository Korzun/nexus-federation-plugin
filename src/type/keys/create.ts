import { addDirective, core } from 'nexus';

import { TypeCreate } from '../../builder';

import { directive, Options as DirectiveOptions } from './directive';
import { object, Options as ObjectOptions } from './object';

export type Options = DirectiveOptions & ObjectOptions;

export const create: TypeCreate<Options> = (builder, options) => {
  /**
   * Adds `@key` directive to the schema
   */
  builder.addOnInstallFunction((t) => {
    t.addType(directive(options));
  });

  /**
   * Adds `keys` to Nexus object definitions and generates the `@key` directive
   * when it is detected in a object definition
   */
  builder.addObjectDefType(object(options));
  builder.addOnObjectDefinitionFunction((block, objectConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (objectConfig.keys) {
      if (!objectConfig.directives) {
        objectConfig.directives = new Array();
      }
      (objectConfig.directives as Array<core.NexusDirectiveUse>).push(
        // @ts-expect-error reflection types have not been generated
        addDirective('key', { fields: objectConfig.keys.join(' ') }),
      );
    }
  });

  return builder;
};
