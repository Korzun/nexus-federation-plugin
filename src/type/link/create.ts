import { TypeCreate } from '../../builder';

import { directive } from './directive';

export const create: TypeCreate = (builder) => {
  builder.addOnInstallFunction((t) => {
    t.addType(directive);
  });

  return builder;
};
