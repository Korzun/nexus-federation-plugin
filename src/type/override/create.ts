import { addDirective, core } from 'nexus';

import { TypeCreate } from '../../builder';

import { directive, Options as DirectiveOptions } from './directive';
import { field, Options as FieldOptions } from './field';

export type Options = DirectiveOptions & FieldOptions;

export const create: TypeCreate<Options> = (builder, options) => {
  /**
   * Adds `@override` directive to the schema
   */
  builder.addOnInstallFunction((t) => {
    t.addType(directive(options));
  });

  /**
   * Adds `override` to Nexus field definitions and generates the `@override`
   * directive when it is detected in a field definition
   */
  builder.addFieldDefType(field(options));
  builder.addOnAddFieldFunction((fieldConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (fieldConfig.override) {
      if (!fieldConfig.directives) {
        fieldConfig.directives = new Array();
      }
      (fieldConfig.directives as Array<core.NexusDirectiveUse>).push(
        // @ts-expect-error reflection types have not been generated
        addDirective('override', { from: fieldConfig.override }),
      );
    }
  });

  return builder;
};
