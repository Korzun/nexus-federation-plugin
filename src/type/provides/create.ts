import { addDirective, core } from 'nexus';

import { TypeCreate } from '../../builder';

import { directive, Options as DirectiveOptions } from './directive';
import { field, Options as FieldOptions } from './field';

export type Options = DirectiveOptions & FieldOptions;

export const create: TypeCreate<Options> = (builder, options) => {
  /**
   * Adds `@provides` directive to the schema
   */
  builder.addOnInstallFunction((t) => {
    t.addType(directive(options));
  });

  /**
   * Adds `provides` to Nexus field definitions and generates the `@provides`
   * directive when it is detected in a field definition
   */
  builder.addFieldDefType(field(options));
  builder.addOnAddFieldFunction((fieldConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (fieldConfig.provides) {
      if (!fieldConfig.directives) {
        fieldConfig.directives = new Array();
      }
      (fieldConfig.directives as Array<core.NexusDirectiveUse>).push(
        // @ts-expect-error reflection types have not been generated
        addDirective('provides', { fields: fieldConfig.provides.join(' ') }),
      );
    }
  });

  return builder;
};
