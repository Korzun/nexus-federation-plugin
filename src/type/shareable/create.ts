import { addDirective, core } from 'nexus';

import { TypeCreate } from '../../builder';

import { directive, Options as DirectiveOptions } from './directive';
import { field, Options as FieldOptions } from './field';
import { object, Options as ObjectOptions } from './object';

export type Options = DirectiveOptions & FieldOptions & ObjectOptions;

export const create: TypeCreate<Options> = (builder, options) => {
  /**
   * Adds `@shareable` directive to the schema
   */
  builder.addOnInstallFunction((t) => {
    t.addType(directive(options));
  });

  /**
   * Adds `shareable` to Nexus object definitions and generates the `@shareable`
   * directive when it is detected in an object definition
   */
  builder.addObjectDefType(object(options));
  builder.addOnObjectDefinitionFunction((t, objectConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (objectConfig.shareable) {
      if (!objectConfig.directives) {
        objectConfig.directives = new Array();
      }
      (objectConfig.directives as Array<core.NexusDirectiveUse>).push(
        addDirective('shareable', {}),
      );
    }
  });

  /**
   * Adds `shareable` to Nexus field definitions and generates the `@shareable`
   * directive when it is detected in a field definition
   */
  builder.addFieldDefType(field(options));
  builder.addOnAddFieldFunction((fieldConfig) => {
    // @ts-expect-error reflection types have not been generated
    if (fieldConfig.shareable) {
      if (!fieldConfig.directives) {
        fieldConfig.directives = new Array();
      }
      (fieldConfig.directives as Array<core.NexusDirectiveUse>).push(
        addDirective('shareable', {}),
      );
    }
  });

  return builder;
};
