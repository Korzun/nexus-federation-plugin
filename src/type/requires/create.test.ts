import { core, objectType } from 'nexus';

import * as testHelper from '../../test';

import { create } from './create';

describe('requires', () => {
  it('is correctly added to schema', () => {
    const schema = testHelper.createSchema(create);
    expect(schema).toMatchSnapshot();
  });
  it('is correctly added to typegen', async () => {
    const types = await testHelper.createTypegen(create);
    expect(types).toMatchSnapshot();
  });

  describe('field configuration', () => {
    describe('is present', () => {
      let testType: core.NexusObjectTypeDef<'User'>;
      beforeEach(() => {
        testType = objectType({
          name: 'User',
          definition: (t) => {
            t.id('id');
            t.string('email');
            // @ts-expect-error reflection types have not been generated
            t.string('name', { requires: ['id', 'email'] });
          },
        });
      });
      it('is correctly added to the schema', () => {
        const schema = testHelper.createSchema(create, {}, testType);
        expect(schema).toMatchSnapshot();
      });
      it('is correctly added to typegen', async () => {
        const typegen = await testHelper.createTypegen(create, {}, testType);
        expect(typegen).toMatchSnapshot();
      });
    });
    describe('is NOT present', () => {
      let testType: core.NexusObjectTypeDef<'User'>;
      beforeEach(() => {
        testType = objectType({
          name: 'User',
          definition: (t) => {
            t.id('id');
            t.string('email');
            t.string('name');
          },
        });
      });
      it('is correctly added to the schema', () => {
        const schema = testHelper.createSchema(create, {}, testType);
        expect(schema).toMatchSnapshot();
      });
      it('is correctly added to typegen', async () => {
        const typegen = await testHelper.createTypegen(create, {}, testType);
        expect(typegen).toMatchSnapshot();
      });
    });
  });
});
