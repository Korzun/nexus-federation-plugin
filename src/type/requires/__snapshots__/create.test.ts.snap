// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`requires field configuration is NOT present is correctly added to the schema 1`] = `
"### This file was generated by Nexus Schema
### Do not make changes to this file directly


\\"\\"\\"

      **Apollo Federation 2.0 Subgraph**
      https://www.apollographql.com/docs/federation/federated-types/federated-directives#requires
      
      Annotates the required input fieldset from a base
      type for a resolver. It is used to develop a query plan
      where the required fields may not be needed by the client,
      but the service may need additional information from other
      services. For example:

      \`\`\`gql
      # extended from the Users service
      extend type User @key(fields: \\"id\\") {
        id: ID! @external
        email: String @external
        reviews: [Review] @requires(fields: \\"email\\")
      }
      \`\`\`

      In this case, the \`Reviews Service\` adds new capabilities
      to the \`User\` type by providing a list of reviews related
      to a user. In order to fetch these reviews, the \`Reviews
      Service\` needs to know the email of the \`User\` from the
      \`Users Service\` in order to look up the reviews. This
      means the reviews field / resolver requires the email field
      from the base \`User\` type.
    
\\"\\"\\"
directive @requires(fields: _FieldSet!) on FIELD_DEFINITION

type Query {
  test: Boolean
}

type User {
  email: String
  id: ID
  name: String
}

\\"\\"\\"

  Represents a set of fields. Grammatically, a field set
  is a selection set minus the braces. This means it can
  represent a single field \\"upc\\", multiple fields \\"id
  countryCode\\", and even nested selection sets \\"id organization
  { id }\\".
  
  [FieldSet - Apollo Federation 2.0 Subgraph Spec](https://www.apollographql.com/docs/federation/subgraph-spec/#scalar-fieldset)
  
\\"\\"\\"
scalar _FieldSet"
`;

exports[`requires field configuration is NOT present is correctly added to typegen 1`] = `
"/**
 * This file was generated by Nexus Schema
 * Do not make changes to this file directly
 */


import type { FederationRequiresField } from \\"@korzun/nexus-federation-plugin/dist/esm/type/requires/field\\"




declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
}

export interface NexusGenScalars {
  String: string
  Int: number
  Float: number
  Boolean: boolean
  ID: string
  _FieldSet: any
}

export interface NexusGenObjects {
  Query: {};
  User: { // root type
    email?: string | null; // String
    id?: string | null; // ID
    name?: string | null; // String
  }
}

export interface NexusGenInterfaces {
}

export interface NexusGenUnions {
}

export type NexusGenRootTypes = NexusGenObjects

export type NexusGenAllTypes = NexusGenRootTypes & NexusGenScalars

export interface NexusGenFieldTypes {
  Query: { // field return type
    test: boolean | null; // Boolean
  }
  User: { // field return type
    email: string | null; // String
    id: string | null; // ID
    name: string | null; // String
  }
}

export interface NexusGenFieldTypeNames {
  Query: { // field return type name
    test: 'Boolean'
  }
  User: { // field return type name
    email: 'String'
    id: 'ID'
    name: 'String'
  }
}

export interface NexusGenArgTypes {
}

export interface NexusGenAbstractTypeMembers {
}

export interface NexusGenTypeInterfaces {
}

export type NexusGenObjectNames = keyof NexusGenObjects;

export type NexusGenInputNames = never;

export type NexusGenEnumNames = never;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = keyof NexusGenScalars;

export type NexusGenUnionNames = never;

export type NexusGenDirectives = \\"requires\\"

export interface NexusGenDirectiveArgs {
  requires: {
    fields:NexusGenScalars['_FieldSet']; // _FieldSet!
  }
}

export type NexusGenObjectsUsingAbstractStrategyIsTypeOf = never;

export type NexusGenAbstractsUsingStrategyResolveType = never;

export type NexusGenFeaturesConfig = {
  abstractTypeStrategies: {
    resolveType: true
    __typename: true
    isTypeOf: false
  }
}

export interface NexusGenTypes {
  context: any;
  inputTypes: NexusGenInputs;
  directives: NexusGenDirectives;
  directiveArgs: NexusGenDirectiveArgs;
  rootTypes: NexusGenRootTypes;
  inputTypeShapes: NexusGenInputs & NexusGenEnums & NexusGenScalars;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  fieldTypeNames: NexusGenFieldTypeNames;
  allTypes: NexusGenAllTypes;
  typeInterfaces: NexusGenTypeInterfaces;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractTypeMembers: NexusGenAbstractTypeMembers;
  objectsUsingAbstractStrategyIsTypeOf: NexusGenObjectsUsingAbstractStrategyIsTypeOf;
  abstractsUsingStrategyResolveType: NexusGenAbstractsUsingStrategyResolveType;
  features: NexusGenFeaturesConfig;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
    
  }
  interface NexusGenPluginInputTypeConfig<TypeName extends string> {
    
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
    /**
     * Annotates the required input fieldset from a base type
     * for a resolver. It is used to develop a query plan where the
     * required fields may not be needed by the client, but the
     * service may need additional information from other services.
     * For example:
     * \`\`\`ts
     * objectType({
     * name: 'User',
     * keyFields: ['id'],
     * definition: (t) => {
     * t.id('id', { external: true });
     * t.nullable.string('email', { external: true });
     * t.list.field('reviews', { type: 'Review', requires: ['email'] });
     * },
     * });
     * \`\`\`
     * In this case, the Reviews service adds new capabilities to the
     * User type by providing a list of reviews related to a user. In
     * order to fetch these reviews, the Reviews service needs to
     * know the email of the User from the Users service in order to
     * look up the reviews. This means the reviews field / resolver
     * requires the email field from the base User type.
     * [Requires - Apollo Federation 2.0 Subgraph Spec](https://www.apollographql.com/docs/federation/federated-types/federated-directives#requires)
     */
    requires?: FederationRequiresField<TypeName>
  }
  interface NexusGenPluginInputFieldConfig<TypeName extends string, FieldName extends string> {
    
  }
  interface NexusGenPluginSchemaConfig {
  }
  interface NexusGenPluginArgConfig {
    
  }
}"
`;

exports[`requires field configuration is present is correctly added to the schema 1`] = `
"### This file was generated by Nexus Schema
### Do not make changes to this file directly


\\"\\"\\"

      **Apollo Federation 2.0 Subgraph**
      https://www.apollographql.com/docs/federation/federated-types/federated-directives#requires
      
      Annotates the required input fieldset from a base
      type for a resolver. It is used to develop a query plan
      where the required fields may not be needed by the client,
      but the service may need additional information from other
      services. For example:

      \`\`\`gql
      # extended from the Users service
      extend type User @key(fields: \\"id\\") {
        id: ID! @external
        email: String @external
        reviews: [Review] @requires(fields: \\"email\\")
      }
      \`\`\`

      In this case, the \`Reviews Service\` adds new capabilities
      to the \`User\` type by providing a list of reviews related
      to a user. In order to fetch these reviews, the \`Reviews
      Service\` needs to know the email of the \`User\` from the
      \`Users Service\` in order to look up the reviews. This
      means the reviews field / resolver requires the email field
      from the base \`User\` type.
    
\\"\\"\\"
directive @requires(fields: _FieldSet!) on FIELD_DEFINITION

type Query {
  test: Boolean
}

type User {
  email: String
  id: ID
  name: String @requires(fields: \\"id email\\")
}

\\"\\"\\"

  Represents a set of fields. Grammatically, a field set
  is a selection set minus the braces. This means it can
  represent a single field \\"upc\\", multiple fields \\"id
  countryCode\\", and even nested selection sets \\"id organization
  { id }\\".
  
  [FieldSet - Apollo Federation 2.0 Subgraph Spec](https://www.apollographql.com/docs/federation/subgraph-spec/#scalar-fieldset)
  
\\"\\"\\"
scalar _FieldSet"
`;

exports[`requires field configuration is present is correctly added to typegen 1`] = `
"/**
 * This file was generated by Nexus Schema
 * Do not make changes to this file directly
 */


import type { FederationRequiresField } from \\"@korzun/nexus-federation-plugin/dist/esm/type/requires/field\\"




declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
}

export interface NexusGenScalars {
  String: string
  Int: number
  Float: number
  Boolean: boolean
  ID: string
  _FieldSet: any
}

export interface NexusGenObjects {
  Query: {};
  User: { // root type
    email?: string | null; // String
    id?: string | null; // ID
    name?: string | null; // String
  }
}

export interface NexusGenInterfaces {
}

export interface NexusGenUnions {
}

export type NexusGenRootTypes = NexusGenObjects

export type NexusGenAllTypes = NexusGenRootTypes & NexusGenScalars

export interface NexusGenFieldTypes {
  Query: { // field return type
    test: boolean | null; // Boolean
  }
  User: { // field return type
    email: string | null; // String
    id: string | null; // ID
    name: string | null; // String
  }
}

export interface NexusGenFieldTypeNames {
  Query: { // field return type name
    test: 'Boolean'
  }
  User: { // field return type name
    email: 'String'
    id: 'ID'
    name: 'String'
  }
}

export interface NexusGenArgTypes {
}

export interface NexusGenAbstractTypeMembers {
}

export interface NexusGenTypeInterfaces {
}

export type NexusGenObjectNames = keyof NexusGenObjects;

export type NexusGenInputNames = never;

export type NexusGenEnumNames = never;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = keyof NexusGenScalars;

export type NexusGenUnionNames = never;

export type NexusGenDirectives = \\"requires\\"

export interface NexusGenDirectiveArgs {
  requires: {
    fields:NexusGenScalars['_FieldSet']; // _FieldSet!
  }
}

export type NexusGenObjectsUsingAbstractStrategyIsTypeOf = never;

export type NexusGenAbstractsUsingStrategyResolveType = never;

export type NexusGenFeaturesConfig = {
  abstractTypeStrategies: {
    resolveType: true
    __typename: true
    isTypeOf: false
  }
}

export interface NexusGenTypes {
  context: any;
  inputTypes: NexusGenInputs;
  directives: NexusGenDirectives;
  directiveArgs: NexusGenDirectiveArgs;
  rootTypes: NexusGenRootTypes;
  inputTypeShapes: NexusGenInputs & NexusGenEnums & NexusGenScalars;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  fieldTypeNames: NexusGenFieldTypeNames;
  allTypes: NexusGenAllTypes;
  typeInterfaces: NexusGenTypeInterfaces;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractTypeMembers: NexusGenAbstractTypeMembers;
  objectsUsingAbstractStrategyIsTypeOf: NexusGenObjectsUsingAbstractStrategyIsTypeOf;
  abstractsUsingStrategyResolveType: NexusGenAbstractsUsingStrategyResolveType;
  features: NexusGenFeaturesConfig;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
    
  }
  interface NexusGenPluginInputTypeConfig<TypeName extends string> {
    
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
    /**
     * Annotates the required input fieldset from a base type
     * for a resolver. It is used to develop a query plan where the
     * required fields may not be needed by the client, but the
     * service may need additional information from other services.
     * For example:
     * \`\`\`ts
     * objectType({
     * name: 'User',
     * keyFields: ['id'],
     * definition: (t) => {
     * t.id('id', { external: true });
     * t.nullable.string('email', { external: true });
     * t.list.field('reviews', { type: 'Review', requires: ['email'] });
     * },
     * });
     * \`\`\`
     * In this case, the Reviews service adds new capabilities to the
     * User type by providing a list of reviews related to a user. In
     * order to fetch these reviews, the Reviews service needs to
     * know the email of the User from the Users service in order to
     * look up the reviews. This means the reviews field / resolver
     * requires the email field from the base User type.
     * [Requires - Apollo Federation 2.0 Subgraph Spec](https://www.apollographql.com/docs/federation/federated-types/federated-directives#requires)
     */
    requires?: FederationRequiresField<TypeName>
  }
  interface NexusGenPluginInputFieldConfig<TypeName extends string, FieldName extends string> {
    
  }
  interface NexusGenPluginSchemaConfig {
  }
  interface NexusGenPluginArgConfig {
    
  }
}"
`;

exports[`requires is correctly added to schema 1`] = `
"### This file was generated by Nexus Schema
### Do not make changes to this file directly


\\"\\"\\"

      **Apollo Federation 2.0 Subgraph**
      https://www.apollographql.com/docs/federation/federated-types/federated-directives#requires
      
      Annotates the required input fieldset from a base
      type for a resolver. It is used to develop a query plan
      where the required fields may not be needed by the client,
      but the service may need additional information from other
      services. For example:

      \`\`\`gql
      # extended from the Users service
      extend type User @key(fields: \\"id\\") {
        id: ID! @external
        email: String @external
        reviews: [Review] @requires(fields: \\"email\\")
      }
      \`\`\`

      In this case, the \`Reviews Service\` adds new capabilities
      to the \`User\` type by providing a list of reviews related
      to a user. In order to fetch these reviews, the \`Reviews
      Service\` needs to know the email of the \`User\` from the
      \`Users Service\` in order to look up the reviews. This
      means the reviews field / resolver requires the email field
      from the base \`User\` type.
    
\\"\\"\\"
directive @requires(fields: _FieldSet!) on FIELD_DEFINITION

type Query {
  test: Boolean
}

\\"\\"\\"

  Represents a set of fields. Grammatically, a field set
  is a selection set minus the braces. This means it can
  represent a single field \\"upc\\", multiple fields \\"id
  countryCode\\", and even nested selection sets \\"id organization
  { id }\\".
  
  [FieldSet - Apollo Federation 2.0 Subgraph Spec](https://www.apollographql.com/docs/federation/subgraph-spec/#scalar-fieldset)
  
\\"\\"\\"
scalar _FieldSet"
`;

exports[`requires is correctly added to typegen 1`] = `
"/**
 * This file was generated by Nexus Schema
 * Do not make changes to this file directly
 */


import type { FederationRequiresField } from \\"@korzun/nexus-federation-plugin/dist/esm/type/requires/field\\"




declare global {
  interface NexusGen extends NexusGenTypes {}
}

export interface NexusGenInputs {
}

export interface NexusGenEnums {
}

export interface NexusGenScalars {
  String: string
  Int: number
  Float: number
  Boolean: boolean
  ID: string
  _FieldSet: any
}

export interface NexusGenObjects {
  Query: {};
}

export interface NexusGenInterfaces {
}

export interface NexusGenUnions {
}

export type NexusGenRootTypes = NexusGenObjects

export type NexusGenAllTypes = NexusGenRootTypes & NexusGenScalars

export interface NexusGenFieldTypes {
  Query: { // field return type
    test: boolean | null; // Boolean
  }
}

export interface NexusGenFieldTypeNames {
  Query: { // field return type name
    test: 'Boolean'
  }
}

export interface NexusGenArgTypes {
}

export interface NexusGenAbstractTypeMembers {
}

export interface NexusGenTypeInterfaces {
}

export type NexusGenObjectNames = keyof NexusGenObjects;

export type NexusGenInputNames = never;

export type NexusGenEnumNames = never;

export type NexusGenInterfaceNames = never;

export type NexusGenScalarNames = keyof NexusGenScalars;

export type NexusGenUnionNames = never;

export type NexusGenDirectives = \\"requires\\"

export interface NexusGenDirectiveArgs {
  requires: {
    fields:NexusGenScalars['_FieldSet']; // _FieldSet!
  }
}

export type NexusGenObjectsUsingAbstractStrategyIsTypeOf = never;

export type NexusGenAbstractsUsingStrategyResolveType = never;

export type NexusGenFeaturesConfig = {
  abstractTypeStrategies: {
    resolveType: true
    __typename: true
    isTypeOf: false
  }
}

export interface NexusGenTypes {
  context: any;
  inputTypes: NexusGenInputs;
  directives: NexusGenDirectives;
  directiveArgs: NexusGenDirectiveArgs;
  rootTypes: NexusGenRootTypes;
  inputTypeShapes: NexusGenInputs & NexusGenEnums & NexusGenScalars;
  argTypes: NexusGenArgTypes;
  fieldTypes: NexusGenFieldTypes;
  fieldTypeNames: NexusGenFieldTypeNames;
  allTypes: NexusGenAllTypes;
  typeInterfaces: NexusGenTypeInterfaces;
  objectNames: NexusGenObjectNames;
  inputNames: NexusGenInputNames;
  enumNames: NexusGenEnumNames;
  interfaceNames: NexusGenInterfaceNames;
  scalarNames: NexusGenScalarNames;
  unionNames: NexusGenUnionNames;
  allInputTypes: NexusGenTypes['inputNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['scalarNames'];
  allOutputTypes: NexusGenTypes['objectNames'] | NexusGenTypes['enumNames'] | NexusGenTypes['unionNames'] | NexusGenTypes['interfaceNames'] | NexusGenTypes['scalarNames'];
  allNamedTypes: NexusGenTypes['allInputTypes'] | NexusGenTypes['allOutputTypes']
  abstractTypes: NexusGenTypes['interfaceNames'] | NexusGenTypes['unionNames'];
  abstractTypeMembers: NexusGenAbstractTypeMembers;
  objectsUsingAbstractStrategyIsTypeOf: NexusGenObjectsUsingAbstractStrategyIsTypeOf;
  abstractsUsingStrategyResolveType: NexusGenAbstractsUsingStrategyResolveType;
  features: NexusGenFeaturesConfig;
}


declare global {
  interface NexusGenPluginTypeConfig<TypeName extends string> {
    
  }
  interface NexusGenPluginInputTypeConfig<TypeName extends string> {
    
  }
  interface NexusGenPluginFieldConfig<TypeName extends string, FieldName extends string> {
    /**
     * Annotates the required input fieldset from a base type
     * for a resolver. It is used to develop a query plan where the
     * required fields may not be needed by the client, but the
     * service may need additional information from other services.
     * For example:
     * \`\`\`ts
     * objectType({
     * name: 'User',
     * keyFields: ['id'],
     * definition: (t) => {
     * t.id('id', { external: true });
     * t.nullable.string('email', { external: true });
     * t.list.field('reviews', { type: 'Review', requires: ['email'] });
     * },
     * });
     * \`\`\`
     * In this case, the Reviews service adds new capabilities to the
     * User type by providing a list of reviews related to a user. In
     * order to fetch these reviews, the Reviews service needs to
     * know the email of the User from the Users service in order to
     * look up the reviews. This means the reviews field / resolver
     * requires the email field from the base User type.
     * [Requires - Apollo Federation 2.0 Subgraph Spec](https://www.apollographql.com/docs/federation/federated-types/federated-directives#requires)
     */
    requires?: FederationRequiresField<TypeName>
  }
  interface NexusGenPluginInputFieldConfig<TypeName extends string, FieldName extends string> {
    
  }
  interface NexusGenPluginSchemaConfig {
  }
  interface NexusGenPluginArgConfig {
    
  }
}"
`;
